c++/src/kj/compat/http.c++

```diff
@@ -3489,6 +3489,358 @@ WebSocketPipe newWebSocketPipe() {
+  // Ensures the extension name is not followed by invalid characters.
+  // Assumes that `extension` starts with `permitted` (caller's responsibility).
+  KJ_REQUIRE(permitted.size() <= extension.size());
+
+  ArrayPtr<const char> tail = extension.slice(permitted.size(), extension.size());
+  stripLeadingSpace(tail);
+
+  return tail.size() == 0;
+}
+
+inline kj::Maybe<CompressionParameters> tryExtractParameters(kj::Vector<String>& configuration,
+    bool isAgreement) {
+  // If the `configuration` is structured correctly and has no invalid parameters/values, we will
+  // return a populated `CompressionParameters` struct.
+  if (configuration.size() == 1) {
+    // Plain `permessage-deflate`.
+    return CompressionParameters{};
+  }
+  auto params = configuration.slice(1, configuration.size());
+  auto keyMaybeValuePairs = toKeysAndVals(params);
+  // Parse parameter strings into parameter[=value] pairs.
+  auto maybeUnverified = populateUnverifiedConfig(keyMaybeValuePairs);
+  KJ_IF_MAYBE(unverified, maybeUnverified) {
+    // Parsing succeeded, i.e. the parameter (`key`) names are valid and we don't have
+    // values for `x_no_context_takeover` parameters (the configuration is structured correctly).
+    // All that's left is to check the `x_max_window_bits` values (if any are present).
+    KJ_IF_MAYBE(validConfig, validateCompressionConfig(kj::mv(*unverified), isAgreement)) {
+      return kj::mv(*validConfig);
+    }
+  }
+  return nullptr;
+}
+
+kj::Vector<CompressionParameters> findValidExtensionOffers(StringPtr offers) {
+  // A function to be called by the client that wants to offer extensions through
+  // `Sec-WebSocket-Extensions`. This function takes the value of the header (a string) and
+  // populates a Vector of all the valid offers.
+  kj::Vector<CompressionParameters> result;
+
+  constexpr auto PERMITTED = "permessage-deflate"_kj;
+  auto extensions = splitPartsWithCondition(offers, ',', PERMITTED);
+
+  for (const auto& offer : extensions) {
+    auto splitOffer = splitPartsWithCondition(offer, ';');
+    if (!isValidExtensionName(splitOffer.front(), PERMITTED)) {
+      continue;
+    }
+    KJ_IF_MAYBE(validated, tryExtractParameters(splitOffer, false)) {
+      // We need to swap the inbound/outbound properties since `tryExtractParameters` thinks we're
+      // parsing as the server (`isAgreement` is false).
+      auto tempCtx = validated->inboundNoContextTakeover;
+      validated->inboundNoContextTakeover = validated->outboundNoContextTakeover;
+      validated->outboundNoContextTakeover = tempCtx;
+      auto tempWindow = validated->inboundMaxWindowBits;
+      validated->inboundMaxWindowBits = validated->outboundMaxWindowBits;
+      validated->outboundMaxWindowBits = tempWindow;
+      result.add(kj::mv(*validated));
+    }
+  }
+
+  return kj::mv(result);
+}
+
+kj::String generatedExtensionRequest(const ArrayPtr<CompressionParameters>& extensions) {
```

Comment:
Should rename this to `generateExtensionRequest` and add it to `http.h` so it can be tested.